### PR TITLE
Introduce materialiser module with DAG logic

### PR DESCRIPTION
**Pull request by sandreae**
## Create materialisation module in `p2panda-rs` with graph logic for:

- converting an array of Messages into an array of Edges
- build a DAG from scratch or from an existing array of Edges
- topologically sort the DAG

### The DAG should:
- not know anything about p2panda types
- just deal with simple String types
- sort topologically choosing leading branches by "highest" string value
 
**fun fact: writing tests for this module will be much easier once https://github.com/p2panda/p2panda/pull/116 is merged**

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header

### CODE REVIEW

File: p2panda-rs/src/materializer/dag.rs
Comment (adzialocha) [inline p2panda-rs/src/materializer/dag.rs:144]: External imports I would move to the top, then crate level imports and then module level.

File: p2panda-rs/src/materializer/utils.rs
Comment (adzialocha) [inline p2panda-rs/src/materializer/utils.rs:68]: `Author` could implement the `Eq` trait, then you would not need to convert them to comparable strings first. Maybe we should make an issue for this (doesn't need to be part of this PR)?

File: p2panda-rs/src/materializer/error.rs
Comment (adzialocha) [inline p2panda-rs/src/materializer/error.rs:11]: I can't see this error being used anywhere (also it does not describe any particular case), can this be removed?

**Review by adzialocha — CHANGES_REQUESTED** [review]
Looking really nice! I have minor comments only regarding dead code and documentation etc.

File: p2panda-rs/src/materializer/utils.rs
Comment (sandreae) [reply p2panda-rs/src/materializer/utils.rs:68]: :+1: I think there already is one? https://github.com/p2panda/p2panda/issues/97

File: p2panda-rs/src/materializer/dag.rs
Comment (sandreae) [reply p2panda-rs/src/materializer/dag.rs:260]: It can, but it requires quite a lot of boilerplate, which would then be duplicated. This will all be gone when we can use `test-utils` so I suggest waiting until both PRs are merged then refactoring.

**Comment by sandreae** [comment]
Think all change requests have been resolved :cold_sweat:

**Review by adzialocha — APPROVED** [review]

**Comment by sandreae** [comment]
I don't want to merge this yet actually.... Not until we've discussed the next steps a bit more, including our implementation of `Instance`. Really we need to rename and update `Operation` (`Message`) to include a `previousOperation` before implementing any of the instance/materialisation logic.

**Comment by adzialocha** [comment]
> I don't want to merge this yet actually.... Not until we've discussed the next steps a bit more, including our implementation of `Instance`. Really we need to rename and update `Operation` (`Message`) to include a `previousOperation` before implementing any of the instance/materialisation logic.

I understand the reasoning but isn't it worth to still merge it? It might take a while until we understood and specified all-things-materialization-logic, and in any case I'm sure some parts implemented in this PR will be reused, even when it comes with potential refactoring.

**Comment by adzialocha** [comment]
For me this PR does NOT implement a materializer, it implements helper method which might be useful for it in the future :+1:

**Comment by sandreae** [comment]
Yes, both points very true, also we wanted to approach it step by step, which this definitely does do. Cool :+1:

File: p2panda-rs/src/materialiser/dag.rs
Comment (cafca) [inline p2panda-rs/src/materialiser/dag.rs:144]: It is merged!

File: p2panda-rs/src/materialiser/dag.rs
Comment (cafca) [inline p2panda-rs/src/materialiser/dag.rs:258]: Should this test be placed with the marshalling function?

**Review by cafca — CHANGES_REQUESTED** [review]
Love the clarity of the implementation. This really feels like it should become its own crate once we know it's stable. I have added some questions and comments. Also check out clippy, I just learned about it and there seem to be some helpful suggestions for the code here.

File: p2panda-rs/src/materialiser/error.rs
Comment (adzialocha) [reply p2panda-rs/src/materialiser/error.rs:11]: Thats maybe more an implicit "error handling pattern" we're using in `p2panda-rs`. We organize errors per module, so the `materialiser` module has its own Error type, the `entry` module has its own Error type and so on .. To know support handling errors coming from the outside they need to be converted. This can be done with the help of the `thiserror` crate and some magic macros. I think this separation helps a lot since Errors in Rust are relatively crazy and this gives clarity where something came from and how the dependencies are.

Errors in Rust is quite a [specific thing](https://www.lpalmieri.com/posts/error-handling-rust/) if you're interested in learning more :crying_cat_face:

File: p2panda-rs/src/materialiser/dag.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/dag.rs]: yeez, nice.

File: p2panda-rs/src/materialiser/dag.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/dag.rs:22]: very nice!

File: p2panda-rs/src/materialiser/dag.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/dag.rs]: Whoop, I'm using it!

File: p2panda-rs/src/materialiser/dag.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/dag.rs]: Yeep

File: p2panda-rs/src/materialiser/dag.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/dag.rs]: moved.

File: p2panda-rs/src/materialiser/utils.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/utils.rs]: Yes, these are for the future. I wrote tests for them :+1:

File: p2panda-rs/src/materialiser/utils.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/utils.rs]: Yes, very good point. I moved these methods into their own files with more explicit names.

File: p2panda-rs/src/materialiser/utils.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/utils.rs]: removed!

File: p2panda-rs/src/materialiser/error.rs
Comment (cafca) [reply p2panda-rs/src/materialiser/error.rs:11]: Thanks!

File: p2panda-rs/src/materialiser/marshall_entries.rs
Comment (cafca) [inline p2panda-rs/src/materialiser/marshall_entries.rs:25]: I think we can't have that in create messages can we? Because we need to make the message, then make an entry from it, and only then we have the id. It's a chicken and egg situation.

If I have this mixed up don't mind this comment, otherwise this comment can be removed I think.

**Review by cafca — CHANGES_REQUESTED** [review]
I have one question here, otherwise this is good to go. I'm sorry about taking so long ... it's year end time.

File: p2panda-rs/src/materialiser/marshall_entries.rs
Comment (sandreae) [reply p2panda-rs/src/materialiser/marshall_entries.rs]: Yes, I think that line may have been left over after some code was removed, not sure what it's alluding to... Thanks! Removed it.

**Pull request merged**
